components/supervisor/pkg/config/gitpod-config-analytics.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package config
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	gitpod "github.com/gitpod-io/gitpod/gitpod-protocol"
+	"github.com/sirupsen/logrus"
+)
+
+type ConfigAnalyzer struct {
+	log    *logrus.Entry
+	report func(field string)
+	delay  time.Duration
+	prev   *gitpod.GitpodConfig
+	timer  *time.Timer
+	mu     sync.RWMutex
+}
+
+func NewConfigAnalyzer(
+	log *logrus.Entry,
+	delay time.Duration,
+	report func(field string),
+	initial *gitpod.GitpodConfig,
+) *ConfigAnalyzer {
+	log.Debug("gitpod config analytics: initialized")
+	return &ConfigAnalyzer{
+		log:    log,
+		delay:  delay,
+		report: report,
+		prev:   initial,
+	}
+}
+
+func (a *ConfigAnalyzer) Analyse(current *gitpod.GitpodConfig) <-chan struct{} {
+	a.log.Debug("gitpod config analytics: schedulling")
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	if a.timer != nil && !a.timer.Stop() {
+		a.log.WithField("cfg", current).Debug("gitpod config analytics: cancelled")
+		<-a.timer.C
+	}
+
+	done := make(chan struct{})
+	a.timer = time.AfterFunc(a.delay, func() {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+		a.process(current)
+		a.timer = nil
+		close(done)
+	})
+	return done
+}
+
+func (a *ConfigAnalyzer) process(current *gitpod.GitpodConfig) {
+	a.log.Debug("gitpod config analytics: processing")
+
+	fields := a.computeFields(a.prev, current)
+	for _, field := range fields {
+		if a.diffByField(a.prev, current, field) {
+			a.report(field)
+		}
+	}
+
+	a.log.WithField("current", current).
+		WithField("prev", a.prev).
+		WithField("fields", fields).
+		Debug("gitpod config analytics: processed")
+
+	a.prev = current
+}
+
+func (a *ConfigAnalyzer) computeFields(configs ...*gitpod.GitpodConfig) []string {
```

Comment:
This function will always returns all keys of `.gitpod.yml` since they are init-ed as a struct, i.e. pointer->nil string->"" int->0.

Or this is as expected?

Reply:
I don't think it is true, at least I don't see it while testing. I can add more tests.

Reply:
This is from the first test case

![image](https://user-images.githubusercontent.com/20944364/181787373-c120d24e-5b0b-4c2e-b58f-e72f3d85b4e7.png)